### PR TITLE
Add a frontend endpoint to retrieve the current sessions cart id

### PR DESCRIPTION
**Pull request by liamoneill**
Allows the completion of https://solvedata.atlassian.net/browse/SOLV-2978

### CODE REVIEW

**Comment by dylan-chong** [comment]
Seem to have gotten it working!

Screenshot 2021-06-04 at 2 34 41 PM

Screenshot 2021-06-04 at 2 34 45 PM

File: Controller/Context/index.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        $this->websiteRepository = $websiteRepository;
+        $this->quoteIdToMaskedQuoteId = $quoteIdToMaskedQuoteId;
+
+        parent::__construct($context);
+    }
+
+    public function execute()
+    {
+        $quoteId = $this->checkoutSession->getQuote()->getId();
+        $quoteIdMasked = empty($quoteId) ? null : $this->getQuoteMaskId((int) $quoteId);
```
Comment (liamoneill) [inline Controller/Context/index.php:45]: Consider wrapping each value in a try/catch block. That way if a single value fails to load we don't fail the entire endpoint.

I'm particularly worried about retrieving the masked quote ID. While I'd be surprised if a masked ID didn't exist for a cart, we've been surprised before 😄.

Comment (liamoneill) [reply Controller/Context/index.php:45]: Nit: Additionally consider moving the branching logic and casting into the `getQuoteMaskId` helper function.

File: Controller/Context/index.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     * @param \Magento\Framework\Json\Helper\Data $jsonHelper
+     */
+    public function __construct(
+        \Magento\Framework\App\Action\Context $context,
+        \Magento\Framework\Json\Helper\Data $jsonHelper,
+        \Magento\Checkout\Model\Session $checkoutSession,
+        \Magento\Store\Model\StoreManagerInterface $storeManager,
+        \Magento\Store\Model\WebsiteRepository $websiteRepository,
+        \Magento\Quote\Model\QuoteIdToMaskedQuoteIdInterface $quoteIdToMaskedQuoteId,
+        Logger $logger
```
Comment (liamoneill) [inline Controller/Context/index.php:30]: Nit: Unused

File: Controller/Context/index.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+use SolveData\Events\Model\Logger;
+
+class Index extends \Magento\Framework\App\Action\Action
+{
+    protected $checkoutSession;
+    protected $jsonHelper;
+    protected $storeManager;
+    protected $logger;
+    protected $websiteRepository;
+    protected $quoteIdToMaskedQuoteId;
```
Comment (liamoneill) [inline Controller/Context/index.php:16]: Nit: I've been super inconsistent on this, but consider making these `private`

File: Controller/Context/index.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * @param \Magento\Framework\App\Action\Context $context
+     * @param \Magento\Checkout\Model\Session $checkoutSession
+     * @param \Magento\Framework\Json\Helper\Data $jsonHelper
```
Comment (liamoneill) [inline Controller/Context/index.php:21]: Nit: update `@params` in doc

File: Controller/Context/index.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        $this->getResponse()->representJson(
+            $this->jsonHelper->jsonEncode($context)
+        );
+    }
+
+    private function getStore()
+    {
+        $websiteId = $this->storeManager->getStore()->getWebsiteId();
+        $website = $this->websiteRepository->getById($websiteId);
+        return $website->getCode();
```
Comment (liamoneill) [inline Controller/Context/index.php:62]:
```suggestion
        $website = $this->storeManager->getWebsite();
        return $website->getCode();
```

File: Controller/Context/index.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SolveData\Events\Controller\Context;
```
Comment (liamoneill) [inline Controller/Context/index.php:5]: I neither love nor hate the `solve/context` path, open to any other naming suggestions?

**Review by liamoneill — COMMENTED** [review]
Looks good to me. Just a few nits, but everything else looks great. Will be good to have this deployed.

File: Controller/Context/index.php (hunk 4 above)
Comment (dylan-chong) [reply Controller/Context/index.php:21]: Why do we even have these, we don't even fill these out?

File: Controller/Context/index.php (hunk 6 above)
Comment (dylan-chong) [reply Controller/Context/index.php:5]: to be honest i think it's good enough. we might add other stuff to it in the future meaning a more specific name would end up being wrong

**Review by dylan-chong — APPROVED** [review]
Review changes done, and self approved :D

**Comment by dylan-chong** [comment]
Gotten the SDK snippet working with this new endpoint

```
// Please keep this snippet in sync with
// https://solvedata.atlassian.net/wiki/spaces/S/pages/1332576276/Magento+SDK+Snippets
// and the snippet on any other Solve Magento customers
solve.addLifecycleHookHandler('beforesend', function (event) {
  /**
   * Grabs order id from the p that contains "Your order # is: 3001234567."
   * Note that Magento users can customise the order success page, so the text
   * and elements can change. In Solve's Magento localdev it is "Your order number is:
   * 000000005.".
   */
  function extractOrderIdFromPage() {
    var textElement = document.querySelector('.checkout-success > p');
    if (textElement === null) return null;

    var match = textElement.innerText.match(/\d+/);
    if (match === null) return null;

    return match[0];
  }

  return fetch('/solve/context')
    .then(function (r) {
      return r.json();
    })
    .then(function (context) {
      var store = context.store;
      var cartId = context.quoteId;

      // If we set the cart_id and have no store, the Solve event processing will
      // error. Stop here and let the store be undefined, so that the Solve
      // commerce linker can hopefully fill it in at some other time.
      if (typeof store !== 'string') return event;
      event.store = store;

      if (typeof cartId === 'string') event.cart_id = cartId;

      if (document.location.pathname.match(/checkout\/onepage\/success/) !== null) {
        var orderId = extractOrderIdFromPage();
        if (typeof orderId === 'string') event.order_id = orderId;
      }

      return event;
    })
    .catch(function (e) {
      console.error('Error fetching /solve/context', e);
      return event;
    });
});
```

File: Controller/Context/index.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    private function getQuoteMaskId($quoteIdString)
+    {
+        try {
+            $quoteId = (int) $quoteIdString;
+            if (empty($quoteId)) return null;
+
+            // Note this returns "" if the mask id doesn't exist (e.g. when
+            // user is signed in). Grr Magento
```
Comment (liamoneill) [inline Controller/Context/index.php:71]:
```suggestion
            // Note this returns an empty string when a mask id does not exist (e.g. when the user is signed in).
```

File: Controller/Context/index.php (hunk 4 above)
Comment (liamoneill) [reply Controller/Context/index.php:20]: Good call out. I've just been following the convention of adding in these type hinting comments from the previous authors.

Feel free to stop following the convention if you want.

**Pull request merged**